### PR TITLE
change(common/lmlayer): always bundle searchTermToKey() with model

### LINKED ISSUE #2958: [LMLayer] Ensure that the default searchTermToKey function is consistent at build and runtime
**Is your feature request related to a problem? Please describe.**
Currently, there are TWO implementations of the "default" `searchTermToKey()` function in the LMLayer. There's one that's used when the Trie gets compiled, and one when the model is actually loaded within the LMLayer proper.

This has already lead to problems! (see #2954)

**Describe the solution you'd like**

Make them consistent! The easy way is to use an advanced software engineering technique called _copying-and-pasting code_. But preferably there will only be one place to store this code.

Another thing we might do is ensure that the `searchTermToKey()` function is included in the model, even when using the default function! It should be compiled as part of the model code along with the rest of the model itself. Therefore, the true implementation of the default `searchTermToKey()` will probably somewhere in the LMLayer compilation code. Note that this code needs to be IE11 compatible 😬

### PR DESCRIPTION
**Pull request by eddieantonio**
This resolves #2958, by *always* bundling the compile-time implementation of the `searchTermToKey` function with the model source code. Therefore, the LMLayer HAS to use the key function used at compile-time!

This is a change to the output of the lexical model compiler. As such, every model compiled with this version of the compiler that did not provide a key function will now contain more code; specifically, the code for the default `searchTermToKey()` function. Incidentally, the default has been copy-pasted from the version created in #2955.

I have a few questions:

 * where should the **definitive** source code of the default `searchTermToKey()` function go? Right now, it's copy-pasted both in the LMLayer and the LMLayer compiler; that said, the version in the LMLayer is a fallback for models that did not specify a key function and were compiled _before_ the introduction of this PR's change.
 * since the default searchTermToKey function is now always bundled, contributing to the compiled model's size, should I try to making this function smaller? There's an "incomplete NFD lookup" table that I think I can shrink by lowercasing the key first before running the lookup.

### CODE REVIEW

**Comment by mcdurdin** [comment]
I would have thought including the normalisation function in the LMLayer and where needed during compilation would have been sufficient -- why do we need to bundle it into every compiled model?

I am also querying the IE11 support here?

**Comment by eddieantonio** [comment]
> I would have thought including the normalisation function in the LMLayer and where needed during compilation would have been sufficient -- why do we need to bundle it into every compiled model?

I am paranoid of the "default" function changing with respect to a compiled model, and thus, breaking the lexical model. Always bundling the key function is one way of prevent backwards-incompatible changes, though I can agree that it is rather dramatic. 

If we don't bundle the default function, then we need to be SUPER careful when making code changes to the default key function; it may break existing lexical models.

> I am also querying the IE11 support here?

So to confirm, IE11 support is no longer a thing in future versions? Then I think there are some scripts in the LMLayer's Karma integration suite that need to be fixed. That also means that I can use `String.prototype.normalize('NFD')` instead of the hack we had before!

**Comment by darcywong00** [comment]
> That also means that I can use String.prototype.normalize('NFD') instead of the hack we had before!

The webView on our oldest supported Android devices (SDK 19, KitKat) might need a polyfill.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/normalize

**Comment by darcywong00** [comment]
Hmm, I tried to test this on Android SDK 19 but get these exceptions trying to load  `nrc.str.sencoten`. I get the same errors on master, so it's something else. @jahorton can you repro?

```
Error occurred when attempting to load dictionary keyman.js:24266
LMLayer.onMessage 
Error: Script error.
    at LMLayerWorker.loadModelFile (blob:null/fccadd54-2684-414d-b939-ed890c822493:1063:22)
    at Object.state.handleMessage (blob:null/fccadd54-2684-414d-b939-ed890c822493:1111:27)
    at LMLayerWorker.onMessage (blob:null/fccadd54-2684-414d-b939-ed890c822493:1004:24) 
```

**Comment by mcdurdin** [comment]
> I am paranoid of the "default" function changing with respect to a compiled model, and thus, breaking the lexical model. Always bundling the key function is one way of prevent backwards-incompatible changes, though I can agree that it is rather dramatic.
> 
> If we don't bundle the default function, then we need to be SUPER careful when making code changes to the default key function; it may break existing lexical models.

The flip side of that of course is that improvements to the default key function then provide no benefit to existing models 😀. I am currently unsure which way to lean on this!

**Comment by mcdurdin** [comment]
> So to confirm, IE11 support is no longer a thing in future versions? Then I think there are some scripts in the LMLayer's Karma integration suite that need to be fixed. That also means that I can use `String.prototype.normalize('NFD')` instead of the hack we had before!

Yes we definitely don't need to worry about IE11 in future versions. However it sounds like Android KitKat/4.4/SDK 19 (ugh multiple version references really guys) might need some TLC. KitKat included Chromium M30 and we need M34 for `normalize`. But read on...

I would like to seriously consider moving us to SDK 21 (Lollipop / 5.0). This release has two significant benefits:

* Chromium is updated by Google rather than device manufacturer (via Google Play). This means that we can move with web standards a lot more freely in the future and don't need to be constrained to years-old baselines.
* We'll definitely have the support we need (M37 included in 5.0 by default; `normalize` added in M34).
* Per [lollipop release notes](https://developer.android.com/about/versions/lollipop#Accessibility): New Input method editor (IME) APIs enable faster switching to other IMEs directly from the input method.

We currently have 2.3% of users on 4.4 according to our dashboard. It seems like there may already be issues for them; they may be better off sticking with an older release of Keyman.

What say you @darcywong00?

**Comment by eddieantonio** [comment]
> > I am paranoid of the "default" function changing with respect to a compiled model, and thus, breaking the lexical model. Always bundling the key function is one way of prevent backwards-incompatible changes, though I can agree that it is rather dramatic.
> > If we don't bundle the default function, then we need to be SUPER careful when making code changes to the default key function; it may break existing lexical models.
> 
> The flip side of that of course is that improvements to the default key function then provide no benefit to existing models 😀. I am currently unsure which way to lean on this!

Yes, but since the key function is used at compile time, we would need to recompile the models and ensure that compiled models specify which version of Keyman they assume. In order to do correct matching, the key function used at compile time and used at runtime _must_ be identical with respect to inputs and outputs — we could change them non-functionally but they must produce the same results.

**Comment by mcdurdin** [comment]
> the key function used at compile time and used at runtime _must_ be identical with respect to inputs and outputs

Gotcha. That makes sense. Well, I'm comfortable with this if we don't need to include the `normalize` polyfill in the compiled models -- because that just makes the models bigger and sadder!

**Comment by mcdurdin** [comment]
In our standup today, we unanimously agreed that Keyman 14.0 would have a baseline version of Android Lollipop 5.0 (SDK 21) with a Chromium baseline of M37. This should help us clean up some of the polyfills that we have been using (e.g. `normalize`).

IE11 is also dead (as is legacy Edge).

[MDN](https://developer.mozilla.org/en-US/) is pretty good at documenting supported versions; e.g. `normalize` appeared in M34 of Chromium.

If I get some time I will go through each platform and document minimum versions that we will test against and support for 14.0. Currently, time is tricky to find!

**Comment by eddieantonio** [comment]
@mcdurdin I got rid of the partial NFD lookup! Now I trust the Keyman platform to polyfill `String.prototype.normalize()`. I understand that's just earlier versions of Android now?

File: common/predictive-text/worker/models/trie-model.ts [hunk 1]
```diff
@@ -414,14 +414,24 @@
   }
 
   /**
-   * Converts word forms in into an indexable form. It does this by converting
-   * the string to uppercase and trying to remove diacritical marks.
+   * Converts wordforms into an indexable form. It does this by converting
```
Comment (mcdurdin) [inline common/predictive-text/worker/models/trie-model.ts:417]:
```suggestion
   * Converts wordforms into an indexable form. It does this by
```

File: common/predictive-text/worker/models/trie-model.ts [hunk 2]
```diff
@@ -414,14 +414,24 @@
...
+   * normalizing the letter case of Latin characters and removing
+   * common diacritical marks.
    *
-   * This is a very naïve implementation, that I've only though to work on
-   * languages that use the Latin script. Even then, some Latin-based
-   * orthographies use code points that, under NFD normalization, do NOT
-   * decompose into an ASCII letter and a combining diacritical mark (e.g.,
-   * SENĆOŦEN).
+   * This is a very naïve implementation, that I only think will work on
+   * some language that use the Latin script. As of 2020-04-08, only
```
Comment (mcdurdin) [inline common/predictive-text/worker/models/trie-model.ts:422]:
```suggestion
   * some languages that use the Latin script. As of 2020-04-08, only
```

File: developer/js/source/lexical-model-compiler/build-trie.ts [hunk 3]
```diff
@@ -201,7 +204,7 @@ namespace Trie {
    * @param keyFunction Function that converts word forms into indexed search keys
    * @returns A JSON-serialiable object that can be given to the TrieModel constructor.
    */
-  export function buildTrie(wordlist: WordList, keyFunction: Wordform2Key = defaultWordform2Key): object {
+  export function buildTrie(wordlist: WordList, keyFunction: Wordform2Key): object {
```
Comment (mcdurdin) [inline developer/js/source/lexical-model-compiler/build-trie.ts:207]: I just realised that this name was a little inconsistent... also needs fixing on line 117 (and elsewhere?)

```suggestion
  export function buildTrie(wordlist: WordList, keyFunction: WordFormToKey): object {
```

File: developer/js/source/lexical-model-compiler/build-trie.ts [hunk 4]
```diff
@@ -380,31 +383,35 @@ namespace Trie {
...
+ * NFD is an easy way to separate a Latin character from its diacritics;
+ * Even then, some Latin-based orthographies use code points that,
+ * under NFD normalization, do NOT decompose into an ASCII letter and a
+ * combining diacritical mark (e.g., SENĆOŦEN).
+ *
+ * Use this only in early iterations of the model. For a production lexical
+ * model, you SHOULD write/generate your own key function, tailored to your
+ * language.
+ */
+export function defaultWordform2Key(wordform: string): string {
```
Comment (mcdurdin) [inline developer/js/source/lexical-model-compiler/build-trie.ts:408]:
```suggestion
export function defaultWordFormToKey(wordform: string): string {
```

Is there any way we can avoid having this function in two places? It seems it should belong in its own file in /common/web/?

File: developer/js/source/lexical-model-compiler/lexical-model-compiler.ts [hunk 5]
```diff
@@ -8,7 +8,7 @@
 import * as ts from "typescript";
 import * as fs from "fs";
 import * as path from "path";
-import { createTrieDataStructure } from "./build-trie";
+import { createTrieDataStructure, defaultWordform2Key } from "./build-trie";
```
Comment (mcdurdin) [inline developer/js/source/lexical-model-compiler/lexical-model-compiler.ts:11]:
```suggestion
import { createTrieDataStructure, defaultWordFormToKey } from "./build-trie";
```

File: developer/js/source/lexical-model-compiler/lexical-model-compiler.ts [hunk 6]
```diff
@@ -71,15 +71,16 @@ export default class LexicalModelCompiler {
         // file, rather than the current working directory.
         let filenames = modelSource.sources.map(filename => path.join(sourcePath, filename));
 
+        // Use the default search term to key function, if left unspecified.
+        let searchTermToKey = modelSource.searchTermToKey || defaultWordform2Key;
```
Comment (mcdurdin) [inline developer/js/source/lexical-model-compiler/lexical-model-compiler.ts:75]:
```suggestion
        let searchTermToKey = modelSource.searchTermToKey || defaultWordFormToKey;
```

File: developer/js/source/lexical-model-compiler/build-trie.ts [hunk 7]
```diff
@@ -380,31 +383,35 @@ namespace Trie {
...
-      // remove diacritical marks.
-      .replace(COMBINING_DIACRITICAL_MARKS, '') as SearchKey;
-  }
+/**
+ * Converts wordforms into an indexable form. It does this by
+ * normalizing the letter case of Latin characters and removing
+ * common diacritical marks.
+ *
+ * This is a very naïve implementation, that I only think will work on
+ * some language that use the Latin script. As of 2020-04-08, only
```
Comment (mcdurdin) [inline developer/js/source/lexical-model-compiler/build-trie.ts:394]:
```suggestion
 * some languages that use the Latin script. As of 2020-04-08, only
```

(I note this comment is not char-for-char identical with the one in trie-model.ts 😉).

File: developer/js/tests/test-default-search-term-to-key.ts [hunk 8]
```diff
@@ -0,0 +1,20 @@
+
+import 'mocha';
+import {assert} from 'chai';
+import { defaultWordform2Key } from '../dist/lexical-model-compiler/build-trie';
```
Comment (mcdurdin) [inline developer/js/tests/test-default-search-term-to-key.ts:4]:
```suggestion
import { defaultWordFormToKey } from '../dist/lexical-model-compiler/build-trie';
```

File: developer/js/tests/test-default-search-term-to-key.ts [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+    // "İstanbul" has a DOTTED-I. This should be lowercased,
+    // then have its dot removed! 
+    assert.equal(defaultWordform2Key('İstanbul'), 'istanbul');
+    assert.equal(defaultWordform2Key('Diyarbakır'), 'diyarbakır');
+
+    // "skýlos" is Greek for dog 🇬🇷🐶
+    // starts with an 's' and ends with an 's'
+    // which are DIFFERENT CHARACTERS in lowercased Greek!
+    assert.equal(defaultWordform2Key('σκύλος'), 'σκυλος');
+    assert.equal(defaultWordform2Key('ΣΚΥΛΟΣ'), 'σκυλος');
```
Comment (mcdurdin) [inline developer/js/tests/test-default-search-term-to-key.ts:18]: Eep. Second worm from the can. 🥫

```suggestion
    assert.equal(defaultWordFormToKey('σκύλος'), 'σκυλος');
    assert.equal(defaultWordFormToKey('ΣΚΥΛΟΣ'), 'σκυλος');
```

Righto. I don't think this is actually what we want, is it? Because `String.toLowerCase()` is [context-sensitive](https://tc39.es/ecma262/#sec-string.prototype.tolowercase), it does _more work_ than we want it to. The differentiation between word-final sigma and sigma is actually unhelpful.

I would think we'd be wanting a char-by-char lower case without context. I imagine we want a result more like:

```suggestion
    assert.equal(defaultWordFormToKey('σκύλος'), 'σκυλοσ');
    assert.equal(defaultWordFormToKey('ΣΚΥΛΟΣ'), 'σκυλοσ');
```

We should consider using NFKD rather than NFD. That transforms chars such as U+01C4 `Ǆ` to U+0044 U+005A `DZ` which is exactly what we want.

More discussion needed?

Ref: https://www.w3.org/International/wiki/Case_folding

File: developer/js/source/lexical-model-compiler/build-trie.ts (hunk 4 above)
Comment (mcdurdin) [reply developer/js/source/lexical-model-compiler/build-trie.ts:408]: I see that this function uses `normalize` but the one in trie-model.ts still uses PARTIAL_NFD_LOOKUP.

File: developer/js/tests/test-default-search-term-to-key.ts [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+import {assert} from 'chai';
+import { defaultWordform2Key } from '../dist/lexical-model-compiler/build-trie';
+
+
+describe('The default searchTermToKey() function', function () {
+  it('should lowercase and THEN normalize', function() {
+    // "İstanbul" has a DOTTED-I. This should be lowercased,
+    // then have its dot removed! 
+    assert.equal(defaultWordform2Key('İstanbul'), 'istanbul');
+    assert.equal(defaultWordform2Key('Diyarbakır'), 'diyarbakır');
```
Comment (mcdurdin) [inline developer/js/tests/test-default-search-term-to-key.ts:12]: This unit test has opened up a can of philosophical worms for me. 🥫

(Side note: please include the Unicode values for the letters in the comment; helps avoid confusion.)

```suggestion
    // "İstanbul" has a U+0130 LATIN CAPITAL LETTER I WITH DOT ABOVE. 
    // This should be lowercased.
    assert.equal(defaultWordFormToKey('İstanbul'), 'istanbul');
    assert.equal(defaultWordFormToKey('Diyarbakır'), 'diyarbakır');
```

This comment is kinda orthographically misleading -- because in Turkish, `İ` lower cases to `i`; the dotless lower-case `ı` upper cases to `I` but we are not removing the dot in the assertion below it 😀. 

So the default normalization here is *not* normalizing U+0131 `ı` to U+0069 `i`, but it is normalizing U+0130 `İ` to U+0069 `i`? That means `'I' == 'İ' == 'i' != 'ı'` which is weird but is what Unicode specifies. But I don't think this is what we want.

There are two possibilities:
1. We want `'I' == 'İ' == 'i' == 'ı'` which makes sense for most Latin language orthographies except Turkic languages. However, we'll never see the U+0130/U+0131 versions except in Turkic languages(*) so this is kinda useless.
2. We are doing a Turkic language model, which means we want `'I' == 'ı', 'İ' == 'i'` but `'I' != 'i'`.

Given this is a _default_ function, we *could* ignore this but the unit test is going to be misleading and unhelpful to implementers either way as it stands!

(*) As far as I know; they may be used in other orthographies but Turkic (e.g. Turkish, Azerbaijani) is the well-known example.

File: developer/js/source/lexical-model-compiler/build-trie.ts (hunk 4 above)
Comment (eddieantonio) [reply developer/js/source/lexical-model-compiler/build-trie.ts:408]: Please see #3014.

Comment (eddieantonio) [reply developer/js/source/lexical-model-compiler/build-trie.ts:408]: >Is there any way we can avoid having this function in two places? It seems it should belong in its own file in /common/web/?

Not... really. Already compiled models in the lexical-models repo _depend_ on a `searchTermToKey` function being bundled with Keyman. Additionally, this differs _slightly_ in the approach of the default provided in the compiler.

Comment (eddieantonio) [reply developer/js/source/lexical-model-compiler/build-trie.ts:408]: It _would_ be easier to share code if we got the Lerna stuff figured out. Then the default search term to key function could be a micro module that is depended upon by both the compiler and the LMLayer. It can be _bundled_ in the LMLayer.

File: developer/js/tests/test-default-search-term-to-key.ts (hunk 10 above)
Comment (eddieantonio) [reply developer/js/tests/test-default-search-term-to-key.ts:12]: So I was half-thinking when I wrote this test. No, I don't think the default function should understand Turkic orthography; it's kind of an edge case in terms of Latin orthographies. The default should be a jack-of-all-orthographies, master-of-none kind of deal. I've updated the tests and comments to address my thoughts.

File: developer/js/tests/test-default-search-term-to-key.ts (hunk 9 above)
Comment (eddieantonio) [reply developer/js/tests/test-default-search-term-to-key.ts:18]: 👍 for NFKD! Again, the change to the default should _only_ happen when to the compiled searchTermToKey function. I like the idea of getting rid of that context-sensitive non-sense from case folding.

I think maybe we'd want to capitalize each letter individually, then lowercase them, again, individually to deal with cases like 'ß' -> 'SS', which should be stored in key form as 'ss'.

Comment (eddieantonio) [reply developer/js/tests/test-default-search-term-to-key.ts]: @mcdurdin, should I leave changing the normalization form (and adding tests for it) to a new PR, or to this PR?

File: developer/js/source/lexical-model-compiler/build-trie.ts (hunk 4 above)
Comment (jahorton) [reply developer/js/source/lexical-model-compiler/build-trie.ts]: I'm already seeing some merit to establishing a few "micro-modules" in other areas too.

File: developer/js/tests/test-default-search-term-to-key.ts (hunk 9 above)
Comment (mcdurdin) [reply developer/js/tests/test-default-search-term-to-key.ts]: I think it belongs as a separate PR. I'm not 100% sure on the round trip on capitalisation; 'ß' -> 'SS' is the only special case I can think of for this at present but there may be others; I haven't gone and checked the Unicode specs yet...

File: developer/js/source/lexical-model-compiler/build-trie.ts (hunk 4 above)
Comment (mcdurdin) [reply developer/js/source/lexical-model-compiler/build-trie.ts]: :+1: for micro modules. As long as we keep the dependency tree sane!
